iotfunctions/anomaly.py

```diff
@@ -325,32 +310,70 @@ def execute(self, df):
                                                         nperseg = self.windowsize, noverlap = self.windowoverlap,
                                                         detrend = False, scaling='spectrum')
 
                 # cut off freqencies too low to fit into the window
-                freqsTSb = (freqsTS > 2/self.windowsize).astype(int)
-                freqsTS = freqsTS * freqsTSb
-                freqsTS[freqsTS == 0] = 1 / self.windowsize
+                frequency_temperatureb = (frequency_temperature > 2/self.windowsize).astype(int)
+                frequency_temperature = frequency_temperature * frequency_temperatureb
+                frequency_temperature[frequency_temperature == 0] = 1 / self.windowsize
+
+                highfrequency_temperature = frequency_temperature.copy()
+                lowfrequency_temperature = frequency_temperature.copy()
+                highfrequency_temperature[highfrequency_temperature <= FrequencySplit] = 0
+                lowfrequency_temperature[lowfrequency_temperature > FrequencySplit] = 0
 
                 # Compute energy = frequency * spectral density over time in decibel
-                ETS = np.log10(np.dot(SxTS.T, freqsTS))
+                try:
+                    lowsignal_energy = np.log10(np.dot(spectral_density_temperature.T, lowfrequency_temperature))
+                    highsignal_energy = np.log10(np.dot(spectral_density_temperature.T, highfrequency_temperature))
+                    signal_energy = np.log10(np.dot(spectral_density_temperature.T, frequency_temperature))
 
-                # compute zscore over the energy
-                ets_zscore = (ETS - ETS.mean())/ETS.std(ddof=0)
-                logger.debug('Spectral z-score max: ' + str(ets_zscore.max()))
+                    # compute the elliptic envelope to exploit Minimum Covariance Determinant estimates
+                    #    standardizing
+                    lowsignal_energy = (lowsignal_energy - lowsignal_energy.mean())/lowsignal_energy.std(ddof=0)
+                    highsignal_energy = (highsignal_energy - highsignal_energy.mean())/highsignal_energy.std(ddof=0)
 
-                # length of timesTS, ETS and ets_zscore is smaller than half the original
-                #   extend it to cover the full original length 
-                Linear = sp.interpolate.interp1d(timesTS, ets_zscore, kind='linear', fill_value='extrapolate')
-                zscoreI = Linear(np.arange(0, temperature.size, 1))
+                    twoDimsignal_energy = np.vstack((lowsignal_energy, highsignal_energy)).T
+                    logger.debug('lowsignal_energy: ' + str(lowsignal_energy) + ', highsignal_energy:' + str(highsignal_energy) + 'input' + str(twoDimsignal_energy))
+
+                    # inliers have a score of 1, outliers -1, and 0 indicates an issue with the data
+                    dfe[self.output_item] = 0.0002
+                    ellEnv = EllipticEnvelope(random_state=0)
```

Comment:
elliptic_envelope ?

Reply:
the name of the method, the assumption is that data is normally distributed. The algo finds a suitable coordinate system for the majority of the data points and looks for the elliptic boundary (with respect to this new coordinate system). Data points outside of the ellipse are outliers